groth16/src/prove.rs

```diff
@@ -74,3 +72,43 @@ pub async fn compute_B<
     let prod = d_msm::<E::G2, _>(&V, &a, &v_pp, net, sid).await?;
     Ok(prod.into_affine().mul(v1))
 }
+
+/// C = (∏{i∈[l+1,m]}(W_i)^a_i)(∏{i∈[0,Q−2]}(U_i)^h_i).A^s.M^r.(∏{i∈[0,m]}(H_i)^a_i)^r
+pub async fn compute_C<
+    E: Pairing<G1Affine = F>,
+    F: FftField + PrimeField,
+    Net: MpcNet,
+>(
+    A: E::G1Affine,
+    M: E::G1,
+    s: F,
+    r: F,
+    (w_pp, W): (PackedSharingParams<E::ScalarField>, Vec<E::G1Affine>),
+    (u_pp, U): (PackedSharingParams<E::ScalarField>, Vec<E::G1Affine>),
+    (h_pp, H): (PackedSharingParams<E::ScalarField>, Vec<E::G1Affine>),
+    a: Vec<E::ScalarField>,
+    h: Vec<E::ScalarField>,
+    net: &mut Net,
+    sid: MultiplexedStreamID,
+) -> Result<E::G1Affine, MpcNetError> {
+    // We use variables (A, M, ∏{i∈[l+1,m]}(W_i)^a_i, ∏{i∈[0,Q−2]}(U_i)h_i, ∏{i∈[0,m]}(H_i)^a_i)
+    // to denote elements in G1. We also assume that all the servers computing the proof
+    // get A, M, s, r and h in the clear and only receive packed shares of the remaining elements.
+
+    // Calculate ∏{i∈[l+1,m]}(W_i)^a_i using dmsm
```

Comment:
Future note: w, u, and h can be parallelized by using a `tokio::join!` against the three futures that calculate w, u, and h, using `MultiplexedStreamId`'s `One`, `Two`, and `Three`.

Reply:
Nice, thanks for the tip. Could you please make that code suggestion?